network/src/peer_store/addr_manager.rs

```diff
@@ -18,14 +18,18 @@ impl AddrManager {
     /// Add an address information to address manager
     pub fn add(&mut self, mut addr_info: AddrInfo) {
         if let Some(key) = multiaddr_to_socketaddr(&addr_info.addr) {
-            if let Some(exists_last_connected_at_ms) = self
-                .get(&addr_info.addr)
-                .map(|addr| addr.last_connected_at_ms)
-            {
+            if let Some(&id) = self.addr_to_id.get(&key) {
+                let (exist_last_connected_at_ms, random_id_pos) = {
+                    let info = self.id_to_info.get(&id).expect("must exists");
+                    (info.last_connected_at_ms, info.random_id_pos)
+                };
                 // Get time earlier than record time, return directly
-                if addr_info.last_connected_at_ms < exists_last_connected_at_ms {
+                if addr_info.last_connected_at_ms < exist_last_connected_at_ms {
                     return;
                 }
+                self.swap_random_id(random_id_pos, self.random_ids.len() - 1);
+                self.random_ids.pop();
+                self.id_to_info.remove(&id);
```

Comment:
I don't know why we need to remove the old id and re-insert it, can we replace the `id_to_info` entry with same id? I think it's simple and easy to read.
```rust
            if let Some(&id) = self.addr_to_id.get(&key) {
                let exist_last_connected_at_ms = self.id_to_info.get(&id).expect("must exists").last_connected_at_ms;
                if addr_info.last_connected_at_ms > exist_last_connected_at_ms {
                    self.id_to_info.insert(id, addr_info);
                }
                return;
            }
```

Reply:
https://github.com/nervosnetwork/ckb/pull/4133/files#diff-d2a5fc6b3c7ac53b73e2b01fa86a9fa42bf03efbfb7df0d7c78dea00bb37d482L51 

We reset the update id just to implement the requirements of this algorithm, I'm not sure what will happen if we don't reset, it may need to ask @jjyr

Reply:
The logic here is that when a duplicate key exists, if the `last_connected_at_ms` satisfies the condition, then replace it with the new value, and this logic has nothing to do with generating randomized positions.

Reply:
In the original logic, the random id corresponding to addr_info has been updated. This random id is something related to the random method. If we don't care about this update, we can directly replace addr info

Reply:
It seems has no side effect, the code delete the old addr and re-insert it, I think we can simply replace the entry.